### PR TITLE
Add RobotPen T9A Support

### PR DESCRIPTION
**Pull request by fangxx3863**
A niche tablet support
Official website -> https://www.robotpen.com/product/wk

### CODE REVIEW

File: OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "Name": "RobotPen T9A",
+  "Specifications": {
+    "Digitizer": {
+      "Width": 300.0,
+      "Height": 210.0,
+      "MaxX": 22784.0,
+      "MaxY": 16896.0
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json:8]: The ratios for Width/Height and MaxX/MaxY do not match. These also doesn't represent a sensible LPI/LPmm, but it's near 2000 LPI.

Did you mean:
```suggestion
      "Width": 289.36,
      "Height": 214.58,
      "MaxX": 22784.0,
      "MaxY": 16896.0
```

File: OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public struct RobotPenTabletReport : ITabletReport
+    {
+        public RobotPenTabletReport(byte[] report)
+        {
+            Raw = report;
+
+            Position = new Vector2
+            {
+                X = Unsafe.ReadUnaligned<ushort>(ref report[6]) | (report[7] << 8),
+                Y = Unsafe.ReadUnaligned<ushort>(ref report[8]) | (report[9] << 8)
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs:16]: The extra fluff here is redundant.

```suggestion
                X = Unsafe.ReadUnaligned<ushort>(ref report[6]),
                Y = Unsafe.ReadUnaligned<ushort>(ref report[8])
```

File: OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenReportParser.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class RobotPenReportParser : IReportParser<IDeviceReport>
+    {
+        public IDeviceReport Parse(byte[] report)
+        {
+            if (report[1].ToString() == "66")
+                return new RobotPenTabletReport(report);
+
+            return new DeviceReport(report);
+        }
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenReportParser.cs:13]: Do *not* use strings to compare numerics ever again.

I'm assuming that non-`0x42` reports are out-of-range. Please do correct me if I'm wrong.

```suggestion
        {
            if (report[1] == 0x42)
                return new RobotPenTabletReport(report);

            return new OutOfRangeReport(report);
        }
```

File: OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            };
+            Pressure = Unsafe.ReadUnaligned<ushort>(ref report[10]);
+
+            Eraser = report[11].IsBitSet(6);
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs:20]: This doesn't make any sense.

Pressure is a `ushort` which spans two bytes, however Eraser is parsed from the same byte that Pressure uses. Are you sure about this? If yes, then apply bit-wise mask to Pressure and avoid conflicting reads.

File: OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+namespace OpenTabletDriver.SystemDrivers.InfoProviders
+{
+    internal class RobotPenDriverInfoDriver : ProcessModuleQueryableDriverInfoProvider
+    {
+        protected override string FriendlyName => "RobotPen";
+
+        protected override string LinuxFriendlyName => "UC Logic";
+
+        protected override string LinuxModuleName => "hid_uclogic";
```
Comment (X9VoiD) [inline OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs:9]: I'd like to see this tested on Linux if this tablet is actually supported by `hid_uclogic`.

File: OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+        protected override string[] WinProcessNames { get; } = new string[]
+        {
+            "TabletDriverCenter",
+            "TabletDriverSetting"
+        };
```
Comment (X9VoiD) [inline OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs:15]: I wanna see proof if these process names are correct.

So basically:

- Install OEM driver.
- Find OEM driver executables.
- Send proof here (preferably a screenshot) of the process names.

File: OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+        protected override string[] Heuristics { get; } = new string[]
+        {
+            "RobotPen"
+        };
```
Comment (X9VoiD) [inline OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs:20]: I also wanna see proof for this.

- Install OEM driver.
- Run `pnputil -e`.
- Send output here.

File: TABLETS.md [hunk 8]
```diff
@@ -87,6 +87,7 @@
 | XP-Pen Star G640S             |     Supported     |
 | XP-Pen Star G960S             |     Supported     |
 | XP-Pen Star G960S Plus        |     Supported     |
+| RobotPen T9A                  |     Supported     |
```
Comment (X9VoiD) [inline TABLETS.md:90]: Alphabetical order please.

**Review by X9VoiD — CHANGES_REQUESTED** [review]

**Comment by fangxx3863** [comment]
I got the real xy from the official SDK 
But the strange thing is that my actual measurements cannot satisfy the aspect ratio 
Do I take an approximation like the picture below?
image
IMAGE 2022-01-26 13:48:55
IMAGE 2022-01-26 13:49:00
IMAGE 2022-01-26 13:49:07

**Comment by X9VoiD** [comment]
- Go to OTD -> Tablet -> Tablet Debugger
- Press "Enable recording"
- *Slowly* hover your pen from top left to bottom right of the physical tablet
- Go to `%localappdata%/OpenTabletDriver`
- Upload `tablet-data.txt`

**Comment by fangxx3863** [comment]
This is `tablet-data.txt` 
[tablet-data.txt](https://github.com/OpenTabletDriver/OpenTabletDriver/files/7942115/tablet-data.txt)
I also write a python script and this picture shows the coordinates where the pen was last detached 
image

File: OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{
+  "Name": "RobotPen T9A",
+  "Specifications": {
+    "Digitizer": {
+      "Width": 282.5,
+      "Height": 208.125,
+      "MaxX": 22600.0,
+      "MaxY": 16650.0
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json:8]:
```suggestion
      "Width": 295.11,
      "Height": 216.69,
      "MaxX": 22656.0,
      "MaxY": 16636.0
```

The MaxX, and MaxY here are final. I assumed that the device has `1950 LPI` resolution to get the new Width and Height, do they look closer to the "truth"?

File: OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    },
+    "Pen": {
+      "MaxPressure": 1023,
+      "Buttons": {
+        "ButtonCount": 3
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json:13]: The pen has three buttons?

Go to Tablet Debugger, then record yourself pressing on the pen buttons three times each. Then upload `tablet-data.txt` here again.

File: OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      }
+    },
+    "AuxiliaryButtons": {
+      "ButtonCount": 1
```
Comment (X9VoiD) [inline OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json:17]: Does the tablet itself (not the pen) only have one button on it?

Go to Tablet Debugger, then record yourself pressing on the tablet buttons. Make sure that the pen is *out of range* of the tablet during recording. Then upload `tablet-data.txt` here again

File: OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+namespace OpenTabletDriver.SystemDrivers.InfoProviders
+{
+    internal class RobotPenDriverInfoDriver : ProcessModuleQueryableDriverInfoProvider
```
Comment (X9VoiD) [inline OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs:3]: Delete this whole class for now.

**Review by X9VoiD — CHANGES_REQUESTED** [review]

File: OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json (hunk 10 above)
Comment (fangxx3863) [reply OpenTabletDriver.Configurations/Configurations/RobotPen/T9A.json:13]: Thanks for the edit, the pen has only one button

File: OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs (hunk 5 above)
Comment (fangxx3863) [reply OpenTabletDriver/SystemDrivers/Providers/RobotPenDriverInfoProvider.cs]: IMAGE 2022-01-27 22:43:45 
The third row

File: OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                X = Unsafe.ReadUnaligned<ushort>(ref report[6]),
+                Y = Unsafe.ReadUnaligned<ushort>(ref report[8])
+            };
+            Pressure = Unsafe.ReadUnaligned<ushort>(ref report[10]);
+
+
+            PenButtons = new bool[]
+            {
+                (report[11] & (1 << 1)) != 0
```
[the remark this answers is not in the document]
Comment (fangxx3863) [reply OpenTabletDriver.Configurations/Parsers/RobotPen/RobotPenTabletReport.cs:23]: YES

**Comment by X9VoiD** [comment]
I still have some doubts with some parts of this PR, but will merge as is and just hotfix when bugs to the report parser are reported in the future.

**Comment by fangxx3863** [comment]
OK, thanks for your reply

**Pull request merged**